### ISSUE TITLE
Add parent id to datavalue query results with multiple occurrences

### ISSUE THREAD

**Issue by ppazos** [ppazos/cabolabs-ehrserver#645]
Structure in a compo:

CLUSTER 0..* at0009
|_ ELEMENT 1..1 at0010
|_ ELEMENT 1..1 at0011

Compo instance 1:

CLUSTER (1)
|_ ELEMENT (1) at0010
|_ ELEMENT (2) at0011
CLUSTER (2)
|_ ELEMENT (3) at0010
|_ ELEMENT (4) at0011
CLUSTER (3)
|_ ELEMENT (5) at0010
|_ ELEMENT (6) at0011

DV query result with group by compo:

[ ELEMENT (1) ... ELEMENT (6) ] 

There is no way to know which ELEMENTs are siblings.

The only solution for that is to add a type of query result that retrieves any intermediate structure, in this case the result needed would be at the CLUSTER level, and the projection is exactly CLUSTER 0009.

That would require to deal with the internal structure of the XML and do correspondent transformations when the output is JSON. This would be better handled by an XML native database if only projections are needed (no conditions).

MySQL supports JSON datatype:
https://dev.mysql.com/doc/refman/5.7/en/json.html

Postgres has some support for both JSON and XML:
https://www.postgresql.org/docs/9.3/static/datatype-xml.html
https://www.postgresql.org/docs/9.4/static/datatype-json.html

**Comment by ppazos** [comment]
The alternative is just to use compo queries and let clients to the work of extracting what they need.

Also can use compo queries with an extractor selector to be applied to the results when the results contain the data. This is not so efficient but will work without much refactoring.

**Comment by ppazos** [comment]
Tested, group by composition, in the result for each compo, groups by path and elements children of different clusters are together:

```
{
	"ed16569a-a810-49d5-9e3d-c7fc4bd1813f": {
		"date": null,
		"cols": [{
				"type": "DV_TEXT",
				"path": "openEHR-EHR-COMPOSITION.test_multiple_cluster_occurrences.v1/data[at0001]/items[at0002]/items[at0003]/value<DV_TEXT>",
				"values": [{
						"value": "texto 1"
					},
					{
						"value": "texto 2"
					},
					{
						"value": "texto 3"
					}
				]
			},
			{
				"type": "DV_COUNT",
				"path": "openEHR-EHR-COMPOSITION.test_multiple_cluster_occurrences.v1/data[at0001]/items[at0002]/items[at0005]/value<DV_COUNT>",
				"values": [{
						"magnitude": 1
					},
					{
						"magnitude": 2
					},
					{
						"magnitude": 3
					}
				]
			}
		]
	}
}
```


A simple solution is let the result be as it is, but add an instance parent id to each value, so values under the same parent can be differentiated and grouped together by processing the results if needed.

This works on any situation with any structure since the path + the parent id is enough to group items at the same level. If the values are at different levels, the path itself is the differentiator, the issue is having multiple occurrences of values at the same level.

**Comment by ppazos** [comment]
To implement the solution described above, this issue should be finished first https://github.com/ppazos/cabolabs-ehrserver/issues/786

**Comment by ppazos** [comment]
MIGRATION: added instanceTemplatePath to DataValueIndex.

**Comment by ppazos** [comment]
Return the instanceTemplatePath on query results so clients can group the data.

Later we will add grouping parameters like "full tree" (group at each level of the path when an instance index is found), "leaves" (only group the last level if there are instance indexes there), "group nth" (group at the nth element in the template path of results, so all descendants from that level will be grouped together, but only at that level, that can be an EVENT in HISTORY.events, and ELEMENTs will be grouped per EVENT).

That will be done on another issue.
